src/pybind/mgr/volumes/module.py

```diff
@@ -200,246 +198,81 @@ def _cmd_fs_subvolumegroup_create(self, inbuf, cmd):
 
     def _cmd_fs_subvolume_create(self, inbuf, cmd):
         """
         :return: a 3-tuple of return code(int), empty string(str), error message (str)
         """
         vol_name = cmd['vol_name']
         sub_name = cmd['sub_name']
-
         size = cmd.get('size', None)
         group_name = cmd.get('group_name', None)
 
-        if not self._volume_exists(vol_name):
-            return -errno.ENOENT, "", \
-                   "Volume '{0}' not found, create it with `ceph fs volume create` " \
-                   "before trying to create subvolumes".format(vol_name)
-
-        # TODO: validate that subvol size fits in volume size
-
-        with SubvolumeClient(self, fs_name=vol_name) as svc:
-            if group_name and not svc.get_group_path(group_name):
-                return -errno.ENOENT, "", \
-                    "Subvolume group '{0}' not found, create it with `ceph fs subvolumegroup create` " \
-                    "before trying to create subvolumes".format(group_name)
-            svp = SubvolumePath(group_name, sub_name)
-            svc.create_subvolume(svp, size)
-
-        return 0, "", ""
+        return self.vc.create_subvolume(vol_name, sub_name, group_name, size)
 
     def _cmd_fs_subvolume_rm(self, inbuf, cmd):
         """
         :return: a 3-tuple of return code(int), empty string(str), error message (str)
         """
         vol_name = cmd['vol_name']
         sub_name = cmd['sub_name']
-
         force = cmd.get('force', False)
         group_name = cmd.get('group_name', None)
 
-        fs = self._volume_get_fs(vol_name)
-        if fs is None:
-            if force:
-                return 0, "", ""
-            else:
-                return -errno.ENOENT, "", \
-                       "Volume '{0}' not found, cannot remove subvolume '{1}'".format(vol_name, sub_name)
-
-        vol_fscid = fs['id']
-
-        with SubvolumeClient(self, fs_name=vol_name) as svc:
-            if group_name and not svc.get_group_path(group_name):
-                if force:
-                    return 0, "", ""
-                else:
-                    return -errno.ENOENT, "", \
-                           "Subvolume group '{0}' not found, cannot remove subvolume '{1}'".format(group_name, sub_name)
-            svp = SubvolumePath(group_name, sub_name)
-            try:
-                svc.delete_subvolume(svp)
-            except cephfs.ObjectNotFound:
-                if force:
-                    return 0, "", ""
-                else:
-                    return -errno.ENOENT, "", \
-                           "Subvolume '{0}' not found, cannot remove it".format(sub_name)
-            svc.purge_subvolume(svp)
-
-        # TODO: purge subvolume asynchronously
-        # self._background_jobs.put(PurgeJob(vol_fscid, svp))
-
-        return 0, "", ""
+        return self.vc.remove_subvolume(vol_name, sub_name, group_name, force)
+
+    def _cmd_fs_subvolume_ls(self, inbuf, cmd):
```

Comment:
Let's not introduce listing of subvolumes in this commit. It should be a separate commit. Last time I introduced listing, Patrick had certain issues around this,
https://github.com/ceph/ceph/pull/27594#discussion_r280285238

Reply:
are you proposing list/get to be in a separate commit or not to have it at all? according to the comment you mention, it seems to me to introduce it later once there is access control for those.

Reply:
I think it should be a separate commit at the least. Then we can check with Patrick.

Reply:
I'll vote for introducing that as a separate pr. @batrick ?

Reply:
separate PR sounds good